### PR TITLE
8279570: IGV: Add source/destination property for load and store nodes with an associated field

### PR DESCRIPTION
**Pull request by chhagedorn**
This patch adds a source/destination property field in the "Properties" view for a load/store node associated with a field in IGV. This simplifies the analysis of graphs when dealing with multiple loads/stores from/to fields.

For each node, we are trying to get the associated `ciField` (`get_field()`). This works fine for direct field load/stores nodes. However, when facing an array access, we need to find the actual node doing the field access to get the `ciField` from. For that purpose, I added an additional search based on pattern matching (`find_source_field_of_array_access()`). It follows the address input of a load/store to find the field `LoadN` (compressed oops) or `LoadP` (non-compressed oops) from which we can get the `ciField`.

Example:
```
iFld = 1; // StoreI 25 (not shown in graph)
iArrFld[2] = 3; // StoreI 67
x = iArrFld2[4][5]; // LoadI 138
```
Screenshot from 2022-01-07 15-12-41

Screenshot from 2022-01-07 15-18-15

- `StoreI 25`: Can directly fetch the field `iFld` with `get_field()`.
- `StoreI 67`: For the store to an array, we follow the address input `AddP 66` -> `46 CastPP` -> `29 DecodeN` -> `28 LoadN` to get the field `iFldArr` from `28 LoadN` (done in `find_source_field_of_array_access()`)
- `LoadI 138`: Same as for `StoreI 67` but we repeat the loop in `find_source_field_of_array_access()` when reaching `LoadI 138` -> ... -> `104 LoadN` to get the field `iArrFld2` eventually from `70 LoadN`.

I additionally ran the following sanity testing with a Hello World program:
```
java -XX:-TieredCompilation -Xcomp -XX:+PrintIdealGraph -XX:PrintIdealGraphLevel=3 -XX:PrintIdealGraphFile=graph.xml HelloWorld.java
```
Thanks,
Christian

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8279570](https://bugs.openjdk.java.net/browse/JDK-8279570): IGV: Add source/destination property for load and store nodes with an associated field

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 315b3806b1051031a19626a9e39f0fbee09ce856
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7048/head:pull/7048` \
`$ git checkout pull/7048`

Update a local copy of the PR: \
`$ git checkout pull/7048` \
`$ git pull https://git.openjdk.java.net/jdk pull/7048/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7048`

View PR using the GUI difftool: \
`$ git pr show -t 7048`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7048.diff">https://git.openjdk.java.net/jdk/pull/7048.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/share/opto/idealGraphPrinter.cpp [hunk 1]
```diff
@@ -628,9 +630,99 @@ void IdealGraphPrinter::visit_node(Node *n, bool edges, VectorSet* temp_set) {
...
+
+// Pattern match on the inputs of 'node' to find load node for the field access.
+Node* IdealGraphPrinter::get_load_node(const Node* node) {
+  Node* load = NULL;
+  Node* addr = node->as_Mem()->in(MemNode::Address);
+  if (addr != NULL && addr->is_AddP()) {
+    Node* base = addr->as_AddP()->base_node();
+    if (base != NULL) {
+      if (base->Opcode() == Op_CastPP) {
+        // CastPP node could have been removed.
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/idealGraphPrinter.cpp:710]: There could also be a `CheckCastPP` or multiple casts, right? I would suggest to use `Node::uncast`.

**Review by TobiHartmann — APPROVED** [review]
Very nice!

File: src/hotspot/share/opto/idealGraphPrinter.cpp (hunk 1 above)
Comment (chhagedorn) [reply src/hotspot/share/opto/idealGraphPrinter.cpp:710]: Good point! I've updated it.

**Review by TobiHartmann — APPROVED** [review]
Looks good.

**Comment by chhagedorn** [comment]
Thanks for your review Tobias!

**Pull request closed without merging**